Added a new column _block_number

SmitaRKulkarni opened the pull request:
Changelog category (leave one):
* New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added a new column _block_number resolves #44532

Implementation:
* _block_number column is not persisted for new blocks.
* For merges if the value is not set then its taken from part info.
* Added MergeTreeSetting `allow_experimental_block_number_column`, only when this is enabled column is persisted. 
* If not persisted, querying this column will return the min block number of part.
Testing:
* Added a test for _block_number column in 02668_column_block_number.sql

SmitaRKulkarni committed:
Added a new column _block_number

Implementation:
* _block_number column is not persisted for new blocks.
* For merges if the value is not set then its taken from part info.
Testing:
* Added a test for _block_number column in 02668_column_block_number.sql

SmitaRKulkarni commented:
@rschu1ze  : Thanks for the review. I will look into the comments. 
This PR is still in progress, I will fix the failing tests (most of the because of virtual column change) and update docs too.

SmitaRKulkarni commented on src/Interpreters/MutationsInterpreter.cpp:
```diff
@@ -475,6 +475,12 @@ static void validateUpdateColumns(
+        /// Dont allow to override value of block number virtual column
+        if (!found && column_name == BlockNumberDescription::COLUMN.name)
+        {
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Update is not supported for virtual column {} ", backQuote(column_name));
```
yes, good point. Will add the check here. thanks

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Addressed review comments and added check for engine, before adding _block_number column to persist

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Updated to persist _block_number column only on horizontal merges

alesapin commented on src/Interpreters/inplaceBlockConversions.cpp:
```diff
@@ -337,9 +338,14 @@ void fillMissingColumns(
-            /// that it is constant everywhere but in some blocks (from other parts) it can be a full column.
-            res_columns[i] = type->createColumnConstWithDefaultValue(num_rows)->convertToFullColumnIfConst();
+            if (requested_column->name == BlockNumberColumn.name)
+                res_columns[i] = type->createColumnConst(num_rows, block_number)->convertToFullColumnIfConst();
```
Not sure, maybe we need to do it only for zero-level parts?

SmitaRKulkarni replied:
`_block_number` column is only empty for zero-level parts

alesapin commented on src/Storages/MergeTree/MergeTask.cpp:
```diff
@@ -877,6 +879,8 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream()
     global_ctx->horizontal_stage_progress = std::make_unique<MergeStageProgress>(
         ctx->column_sizes ? ctx->column_sizes->keyColumnsWeight() : 1.0);
 
+    global_ctx->merging_column_names.emplace_back(BlockNumberColumn.name);
```
Merging columns are PK columns: https://github.com/ClickHouse/ClickHouse/blob/a50e741bea349775154278fb39a5afd290174ee8/src/Storages/MergeTree/MergeTask.cpp#L84-L96
gathering columns are non sorting column. So for horizontal merge BlockNumberColumn should be merging_column and for vertical merge it should be gathering_columns: https://github.com/ClickHouse/ClickHouse/blob/a50e741bea349775154278fb39a5afd290174ee8/src/Storages/MergeTree/MergeTask.cpp#L278-L303

SmitaRKulkarni committed:
Updated to use gathering_column for vertical merge

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Fixed build issues

SmitaRKulkarni committed:
Removed duplicate includes

SmitaRKulkarni committed:
Fixed fast test fails and removed _block_number for ttl and deduplicate

SmitaRKulkarni committed:
Updated check for adding merging columns and fixed duplicate includes

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #47532 from ClickHouse/ClickHouse:44532_Persistent_virtual_column__block_number -->
This is an automated comment for commit ef5f2cf7ab960377ede4cf2f5e62418a8afab54d with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/47532/ef5f2cf7ab960377ede4cf2f5e62418a8afab54d/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

SmitaRKulkarni committed:
Fixed build issues

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Fixed deduplicate

SmitaRKulkarni committed:
Fixed deduplicate and removed support for tables with projections

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Updated gathering columns only when no projections

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Removed checking physical column for _block_number when dynamic columns are processed

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Fixed issue with compact parts

SmitaRKulkarni committed:
Updated adding _block_number column only when its not added by previous merge

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Fixed build issue

SmitaRKulkarni committed:
Fixed issue with lightweight delete and updated with new changes for available_columns_set

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Updated to include column _block_number to merging_column_names

SmitaRKulkarni committed:
Added projection check

SmitaRKulkarni committed:
Removed adding _block_number column in MutateTask

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Fixed build

SmitaRKulkarni committed:
Updated to add _block_number column for mutate to fix issue with modify & ttl

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Added setting for enabling persisting of _block_number column and added projections check

SmitaRKulkarni committed:
Fixes for alter command and added check for projections in StorageSnapshot

SmitaRKulkarni committed:
Added _block_number column in mutateOriginalPartAndPrepareProjections

SmitaRKulkarni committed:
Removed _block_number column addition for mutations and fixed lightweight_delete_and_ttl with _block_number issue

SmitaRKulkarni committed:
Updated adding _block_number column in MergeTask to avoid inconsistency between columns.txt & serialization.json

SmitaRKulkarni committed:
Added check for ttl which was removed in previous commit

SmitaRKulkarni committed:
Updated to set serialization info if not found in loaded columns for _block_number column

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Added docs for _block_number columns and setting

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Fixed serialization issue and fixed tests

robot-clickhouse committed:
Automatic style fix

SmitaRKulkarni committed:
Fixed tests with allow persisting block_number

SmitaRKulkarni committed:
Updated integration & stateless test with block_number file size and count

SmitaRKulkarni committed:
Fixed issue with 02490_replacing_merge_tree_is_deleted_column

SmitaRKulkarni commented on src/Interpreters/MutationsInterpreter.cpp:
```diff
@@ -475,6 +475,12 @@ static void validateUpdateColumns(
             found = true;
         }
 
+        /// Dont allow to override value of block number virtual column
```
yes, looks like it  (called from `MutationsInterpreter::prepare`)
https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/MutationsInterpreter.cpp#L533C2

SmitaRKulkarni commented on tests/queries/0_stateless/02668_column_block_number.sql:
```diff
@@ -0,0 +1,33 @@
+DROP TABLE IF EXISTS test;
+CREATE TABLE test (id UInt32, a UInt32) ENGINE = MergeTree ORDER BY id;
+
+-- SYSTEM STOP MERGES test;
```
addressed it, adding setting `allow_experimental_block_number`

SmitaRKulkarni committed:
Set allow_experimental_block_number_column default to false, updated tests and docs. Moved allow_experimental_block_number_column to MergeTreeSettings

SmitaRKulkarni commented on src/Interpreters/inplaceBlockConversions.cpp:
```diff
@@ -333,7 +334,13 @@ void fillMissingColumns(
             /// We must turn a constant column into a full column because the interpreter could infer
             /// that it is constant everywhere but in some blocks (from other parts) it can be a full column.
-            res_columns[i] = type->createColumnConstWithDefaultValue(num_rows)->convertToFullColumnIfConst();
+
```
addressed it

SmitaRKulkarni commented on src/Storages/MergeTree/IMergeTreeDataPart.h:
```diff
@@ -439,6 +440,8 @@ class IMergeTreeDataPart : public std::enable_shared_from_this<IMergeTreeDataPar
     /// True if here is lightweight deleted mask file in part.
     bool hasLightweightDelete() const { return columns.contains(LightweightDeleteDescription::FILTER_COLUMN.name); }
 
+    bool hasBlockNumber() const { return columns.contains(BlockNumberDescription::COLUMN.name); }
```
Addressed the first comment. and removed the function as its not needed

SmitaRKulkarni commented on src/Storages/BlockNumberDescription.cpp:
```diff
@@ -0,0 +1,9 @@
+namespace DB
+{
+
+const NameAndTypePair BlockNumberDescription::COLUMN {"_block_number", std::make_shared<DataTypeUInt64>()};
```
I looked into this, in couple of places we call it as _block_number as well

SmitaRKulkarni commented on src/Storages/BlockNumberDescription.h:
```diff
@@ -0,0 +1,13 @@
+namespace DB
+{
+
+struct BlockNumberDescription
```
ya I had kept it same as LightweightDeleteDescription::FILTER_COLUMN, but atleast for BlockNumberColumn I think we can just keep the NameAndTypePair. 
addressed it.

SmitaRKulkarni commented on src/Storages/BlockNumberDescription.h:
```diff
@@ -0,0 +1,13 @@
+#pragma once
+#include <Core/NamesAndTypes.h>
+#include "Storages/TTLDescription.h"
```
addressed it

SmitaRKulkarni commented on tests/queries/0_stateless/02668_column_block_number.sql:
```diff
@@ -0,0 +1,33 @@
+DROP TABLE IF EXISTS test;
```
addressed it

SmitaRKulkarni committed:
Fixed integration tests and fixed includes

robot-clickhouse committed:
Automatic style fix

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Updated to use delta codec for block number column

alesapin commented on src/Storages/MergeTree/MergeTask.cpp:
```diff
@@ -996,6 +1016,17 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream()
 
     if (global_ctx->deduplicate)
     {
+        if (data_settings->allow_experimental_block_number_column
```
I don't completely understand this branch, need comment.

alesapin commented on src/Storages/MergeTree/MergeTreeDataPartCompact.cpp:
```diff
@@ -64,6 +65,9 @@ IMergeTreeDataPart::MergeTreeWriterPtr MergeTreeDataPartCompact::getWriter(
         { return *getColumnPosition(lhs.name) < *getColumnPosition(rhs.name); });
 
+    if (columns_list.contains(BlockNumberColumn.name) && !ordered_columns_list.contains(BlockNumberColumn.name))
+        ordered_columns_list.emplace_back(BlockNumberColumn);
```
Ok, but need a comment

SmitaRKulkarni replied:
addressed it

alesapin commented on tests/queries/0_stateless/02668_column_block_number.sql:
```diff
@@ -0,0 +1,32 @@
+INSERT INTO test(id,a) VALUES (7,7),(8,8),(9,9);
+
+SELECT '*** AFTER MUTATION AFTER MERGE , NEW BLOCK ***';
+SELECT *,_block_number,_part from test ORDER BY id;
```
We need a test with vertical merge.

SmitaRKulkarni replied:
added test with vertical merge

alesapin commented on src/Storages/MergeTree/MergeTask.cpp:
```diff
@@ -889,6 +901,14 @@ void MergeTask::ExecuteAndFinalizeHorizontalPart::createMergedStream()
         ctx->column_sizes ? ctx->column_sizes->keyColumnsWeight() : 1.0);
 
+    if (data_settings->allow_experimental_block_number_column
+        && !ctx->need_remove_expired_values && global_ctx->metadata_snapshot->getProjections().empty()
```
Why we shouldn't have any projections? What if someone will add projection after `block_number` already exists?

SmitaRKulkarni replied:
works with projection, updated check

alesapin commented on tests/queries/0_stateless/02668_column_block_number.sql:
```diff
@@ -0,0 +1,32 @@
+DROP TABLE IF EXISTS test;
+
+CREATE TABLE test (id UInt32, a UInt32) ENGINE = MergeTree ORDER BY id SETTINGS allow_experimental_block_number_column = true;
```
We need test with projections

SmitaRKulkarni replied:
added test with projections

alesapin commented on src/Storages/MergeTree/MergeTask.cpp:
```diff
@@ -251,12 +252,23 @@ bool MergeTask::ExecuteAndFinalizeHorizontalPart::prepare()
         ctx->need_remove_expired_values = true;
 
+    if (global_ctx->data->getSettings()->allow_experimental_block_number_column
+        && !ctx->need_remove_expired_values && global_ctx->metadata_snapshot->getProjections().empty()
```
Why `!ctx->need_remove_expired_values`? It's relate to TTL and it will just filter our some rows after merge. Should be ok with block number columns.

SmitaRKulkarni replied:
ya you are right, fixed it. Updated to not persist `_block_number` column if `groupByTTL` exists

alesapin commented on src/Storages/ColumnsDescription.cpp:
```diff
@@ -714,6 +718,12 @@ bool ColumnsDescription::hasCompressionCodec(const String & column_name) const
 
 CompressionCodecPtr ColumnsDescription::getCodecOrDefault(const String & column_name, CompressionCodecPtr default_codec) const
 {
+    if (column_name == BlockNumberColumn.name)
```
It's slightly strange that we add this logic into general structure `ColumnsDescription`. We can do it, but I'd prefer to add assertion here (column_name != BlockNumberColumn.name) and add special method for BlockNumber column.

SmitaRKulkarni replied:
addressed it

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Enabled projections & ttl, turned on allow_experimental_block_number_column

SmitaRKulkarni committed:
Disable persisting _block_number column when groupByTTL is used, updated tests to work with persisting _block_number column

SmitaRKulkarni committed:
Updated compress_marks test

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Added test tests/queries/0_stateless/02861_replacing_merge_tree_with_block_number_column.sql

SmitaRKulkarni commented:
I have added a test `02861_replacing_merge_tree_with_block_number_column.sql` , when I run it I see this error 
`Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Number of rows in source parts (12) excluding filtered rows (0) differs from number of bytes written to rows_sources file (10). It is a bug.. (LOGICAL_ERROR)`

This is only seen when vertical merge is used for parts all_1_2_1 & all_3_3_0, and OPTIMIZE is used with CLEANUP. (Without CLEANUP even when vertical merge it works fine). 
I suspect that in ReplacingSortedAlgorithm we replace new values for the old primary key, and since this column value is not present in the new values row, we see an issue.

@alesapin  : Please let me know if you have any ideas what could be happening

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Disabled block_number column for ReplacingMergeTree engine

SmitaRKulkarni committed:
Added test for vertical merge with block number column and fixed integration test to work with default block number column enabled

SmitaRKulkarni commented:
Currently disabled persisting column for ReplacingMergeTree

SmitaRKulkarni committed:
Added tests and addressed review comments

SmitaRKulkarni committed:
Fixed deduplicate condition and added comment

SmitaRKulkarni committed:
Fixed test 02362_part_log_merge_algorithm

alesapin committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

alesapin committed:
Merge branch 'fix_replacing_final_error' into 44532_Persistent_virtual_column__block_number

alesapin committed:
Fix for replacing

alesapin commented on src/Storages/StorageLog.cpp:
```diff
@@ -452,10 +455,20 @@ void LogSink::writeData(const NameAndTypePair & name_and_type, const IColumn & c
+            if (name_and_type.name == BlockNumberColumn.name)
+            {
+                auto data_bytes_size = BlockNumberColumn.type->getSizeOfValueInMemory();
+                compression = getCompressionCodecDelta(data_bytes_size);
```
Compression codec delta without additional compression on top of it doesn't do anything. We need something like `Delta + ZSTD` or `Delta + LZ4`

SmitaRKulkarni committed:
Updated compression codec for BlockNumberColumn

SmitaRKulkarni commented on src/Interpreters/MutationsInterpreter.cpp:
```diff
@@ -1055,6 +1065,18 @@ struct VirtualColumns
+                if (!part->getColumns().contains(BlockNumberColumn.name))
+                {
+                    ColumnWithTypeAndName mask_column;
+                    mask_column.type = BlockNumberColumn.type;
```
addressed it

SmitaRKulkarni commented on src/Storages/BlockNumberColumn.cpp:
```diff
@@ -0,0 +1,23 @@
+
+CompressionCodecPtr getCompressionCodecDelta(UInt8 delta_bytes_size);
+
+CompressionCodecPtr getCompressionCodecForBlockNumberColumn()
+{
+    std::vector <CompressionCodecPtr> codecs;
+    codecs.reserve(2);
+    auto data_bytes_size = BlockNumberColumn::type->getSizeOfValueInMemory();
+    codecs.emplace_back(getCompressionCodecDelta(data_bytes_size));
+    codecs.emplace_back(CompressionCodecFactory::instance().get("LZ4", {}));
+    return std::make_shared<CompressionCodecMultiple>(codecs);
+}
```
@alesapin : I have added setting compression codec using CompressionCodecMultiple . Let me know if there is a better way to add this.

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number

SmitaRKulkarni committed:
Updated setting block_number column to avoid adding serialization info

SmitaRKulkarni committed:
Moved back to add serialization info

SmitaRKulkarni committed:
- Updated with new way of reading MergeTree data
- Disabled allow_experimental_block_number_column by default and fixed tests accordingly

SmitaRKulkarni committed:
Merge branch 'master' into 44532_Persistent_virtual_column__block_number
